### PR TITLE
Raknet Refactor

### PR DESCRIPTION
**Pull request by SupremeMortal**
This PR improves performance by removing as much object allocation as possible and other optimisations for caching sent datagrams etc.
The version has been bumped to 2.0.0 due to the major API changes that come with this PR.

### CODE REVIEW

**Comment by NycuRO** [comment]
Congrats!

File: raknet/src/main/java/com/nukkitx/network/raknet/RakNetUtils.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import java.util.Arrays;
+
+@UtilityClass
+public class RakNetUtils {
+
+    /*
+        Packet IDs
+     */
+
+    public static final InetSocketAddress LOOPBACK_V4 = new InetSocketAddress(Inet4Address.getLoopbackAddress(), 19132);
```
Comment (NycuRO) [inline raknet/src/main/java/com/nukkitx/network/raknet/RakNetUtils.java]: why this is hardcoded to 19132?

File: raknet/src/main/java/com/nukkitx/network/raknet/util/BitQueue.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        this.queue = this.newArray(this.capacity + 1);
+        this.head = 0;
+        this.tail = 0;
+    }
+
+    public void resize(int newCapacity) {
+        int size = this.size();
+        if (newCapacity < size)
+            throw new IndexOutOfBoundsException("Resizing would lose data");
+        newCapacity++;
```
Comment (NycuRO) [inline raknet/src/main/java/com/nukkitx/network/raknet/util/BitQueue.java]: you sure this cannot be implemented in other way?

Comment (DaMatrix) [reply raknet/src/main/java/com/nukkitx/network/raknet/util/BitQueue.java]: does it need to be?

Comment (NycuRO) [reply raknet/src/main/java/com/nukkitx/network/raknet/util/BitQueue.java]: Maybe? It's kinda too ew.

File: raknet/src/main/java/com/nukkitx/network/raknet/RakNetUtils.java (hunk 1 above)
Comment (SupremeMortal) [reply raknet/src/main/java/com/nukkitx/network/raknet/RakNetUtils.java]: It doesn't matter what port it's at.

**Pull request merged**